File: earnings-call-analyzer/eca/analyze.py
"""Analysis engine: structured, cited extraction from a transcript.

Two backends behind one interface:
  - mock  : deterministic heuristic extractor (no API key, no network). Doubles as the
            baseline the eval harness scores real runs against.
  - claude: real analysis via the Anthropic API when ANTHROPIC_API_KEY is set.

Both return the same schema (see prompts.SECTIONS), with every item carrying real
line citations so the grounding layer can verify them.
"""

from __future__ import annotations

import json
import os
import re

from . import prompts
from .ingest import Transcript

# --- Small Loughran-McDonald-style sentiment lexicon (baseline, not novel) ----------
# A production build would load the full LM master dictionary; this compact subset keeps
# the mock deterministic and dependency-free while illustrating the method.
LM_POSITIVE = {
    "growth", "grew", "strong", "record", "improved", "improvement", "gain", "gains",
    "profit", "profitable", "outperform", "beat", "exceeded", "confident", "momentum",
    "accelerate", "expansion", "robust", "healthy", "opportunity", "progress",
}
LM_NEGATIVE = {
    "decline", "declined", "weak", "weakness", "loss", "losses", "headwind", "headwinds",
    "pressure", "uncertain", "uncertainty", "challenge", "challenging", "slowdown",
    "miss", "missed", "risk", "risks", "soft", "softness", "adverse", "litigation",
}

_MONEY = re.compile(r"\$\s?\d[\d,]*(?:\.\d+)?\s?(?:billion|million|trillion|bn|mn|b|m)?", re.I)
_PERCENT = re.compile(r"\d[\d,]*(?:\.\d+)?\s?%")
_NUMBER = re.compile(r"\d[\d,]*(?:\.\d+)?")
# A guided range like "$4.4 to $4.6 billion" or "31 to 32%".
_MONEY_RANGE = re.compile(
    r"\$\s?\d[\d,]*(?:\.\d+)?\s*(?:to|-|–|and)\s*\$?\s?\d[\d,]*(?:\.\d+)?\s?"
    r"(?:billion|million|trillion|bn|mn|b|m)?",
    re.I,
)
_PERCENT_RANGE = re.compile(r"\d[\d,]*(?:\.\d+)?\s*(?:to|-|–)\s*\d[\d,]*(?:\.\d+)?\s?%")

_METRIC_KEYWORDS = [
    ("EPS", re.compile(r"\b(eps|earnings per share|diluted (?:eps|earnings))\b", re.I)),
    ("Revenue", re.compile(r"\b(revenue|net sales|total sales|top line)\b", re.I)),
    ("Net income", re.compile(r"\b(net income|net earnings|net loss)\b", re.I)),
    ("Operating margin", re.compile(r"\b(operating margin|gross margin|operating income)\b", re.I)),
    ("Free cash flow", re.compile(r"\b(free cash flow|operating cash flow|cash flow)\b", re.I)),
    ("Guidance", re.compile(r"\b(guidance|outlook|expect|forecast|full[- ]year|fiscal)\b", re.I)),
    ("Buyback / dividend", re.compile(r"\b(buyback|repurchase|dividend|return of capital)\b", re.I)),
]

_GUIDANCE_RE = re.compile(r"\b(guidance|outlook|expect|anticipate|forecast|full[- ]year|next quarter|fiscal \d{4})\b", re.I)
_RISK_RE = re.compile(r"\b(risk|headwind|uncertain|challenge|pressure|macro|foreign exchange|fx|slowdown|softness)\b", re.I)
_QA_MARKER = re.compile(r"\b(q&a|question[- ]and[- ]answer|first question|analyst)\b", re.I)


def analyze(
    transcript: Transcript,
    company_hint: str = "",
    quarter_hint: str = "",
    backend: str = "mock",
    model: str | None = None,
) -> dict:
    if backend == "claude":
        return _analyze_claude(transcript, company_hint, quarter_hint, model)
    return _analyze_mock(transcript, company_hint, quarter_hint)


# --- Mock backend --------------------------------------------------------------------

def _analyze_mock(transcript: Transcript, company_hint: str, quarter_hint: str) -> dict:
    lines = transcript.lines
    company = company_hint or _guess_company(lines)
    quarter = quarter_hint or _guess_quarter(lines)

    key_metrics = _extract_metrics(lines)
    guidance = _extract_matches(lines, _GUIDANCE_RE, limit=4)
    risks = _extract_matches(lines, _RISK_RE, limit=4)
    qa = _extract_qa(lines)
    tone = _extract_tone(lines)
    quotes = _extract_quotes(lines)

    # Executive summary: top metric lines + the sharpest tone read.
    summary: list[dict] = []
    for m in key_metrics[:3]:
        summary.append({"text": f"{m['metric']} reported at {m['value']}.", "citations": m["citations"]})
    if tone:
        summary.append(tone[0])

    return {
        "company": company,
        "quarter": quarter,
        "executive_summary": summary,
        "guidance": guidance,
        "key_metrics": key_metrics,
        "management_tone": tone,
        "risks": risks,
        "qa_highlights": qa,
        "notable_quotes": quotes,
    }


_COMPANY_RE = re.compile(
    r"((?:[A-Z][A-Za-z0-9.&'’]+,?\s+){1,5}"
    r"(?:Corporation|Incorporated|Company|Holdings|Group|Corp|Inc|Ltd|PLC|Co|N\.V)\.?)"
)


def _guess_company(lines: list[str]) -> str:
    for text in lines[:12]:
        m = _COMPANY_RE.search(text)
        if m:
            return m.group(1).strip().rstrip(",")
    return "Unknown Company"


def _guess_quarter(lines: list[str]) -> str:
    for text in lines[:20]:
        m = re.search(r"\b(Q[1-4]|first|second|third|fourth)[- ]?(?:quarter)?[, ]*(?:of\s*)?(?:fiscal\s*)?(20\d{2})\b", text, re.I)
        if m:
            return f"{m.group(1).upper()} {m.group(2)}"
    return "Unknown Quarter"


def _extract_metrics(lines: list[str]) -> list[dict]:
    metrics: list[dict] = []
    seen: set[str] = set()
    for i, text in enumerate(lines, start=1):
        if not (_MONEY.search(text) or _PERCENT.search(text)):
            continue
        for label, kw in _METRIC_KEYWORDS:
            if label in seen:
                continue
            if kw.search(text):
                value = _pick_value(text)
                if value:
                    metrics.append({"metric": label, "value": value, "citations": [i]})
                    seen.add(label)
                break
    return metrics


def _pick_value(text: str) -> str:
    money_range = _MONEY_RANGE.search(text)
    if money_range:
        return re.sub(r"\s+", " ", money_range.group(0).strip())
    pct_range = _PERCENT_RANGE.search(text)
    if pct_range:
        return re.sub(r"\s+", " ", pct_range.group(0).strip())
    money = _MONEY.search(text)
    pct = _PERCENT.search(text)
    if money and pct:
        # Report both when a line pairs an absolute with a growth rate.
        return f"{money.group(0).strip()} ({pct.group(0).strip()})"
    if money:
        return money.group(0).strip()
    if pct:
        return pct.group(0).strip()
    num = _NUMBER.search(text)
    return num.group(0) if num else ""


def _extract_matches(lines: list[str], pattern: re.Pattern, limit: int) -> list[dict]:
    out: list[dict] = []
    for i, text in enumerate(lines, start=1):
        if pattern.search(text):
            out.append({"text": _clip(text), "citations": [i]})
            if len(out) >= limit:
                break
    return out


def _extract_qa(lines: list[str]) -> list[dict]:
    start = None
    for i, text in enumerate(lines, start=1):
        if _QA_MARKER.search(text):
            start = i
            break
    if start is None:
        return []
    out: list[dict] = []
    for i in range(start, len(lines) + 1):
        text = lines[i - 1]
        if re.match(r"^(analyst|q:|question)", text, re.I) or (_PERCENT.search(text) or _MONEY.search(text)):
            out.append({"text": _clip(text), "citations": [i]})
            if len(out) >= 4:
                break
    return out


def _extract_tone(lines: list[str]) -> list[dict]:
    pos_hits: list[int] = []
    neg_hits: list[int] = []
    for i, text in enumerate(lines, start=1):
        words = set(re.findall(r"[a-z]+", text.lower()))
        if words & LM_POSITIVE:
            pos_hits.append(i)
        if words & LM_NEGATIVE:
            neg_hits.append(i)
    total = len(pos_hits) + len(neg_hits)
    if total == 0:
        return []
    net = (len(pos_hits) - len(neg_hits)) / total
    lean = "net positive" if net > 0.15 else "net negative" if net < -0.15 else "balanced"
    cites = (pos_hits[:2] + neg_hits[:2]) or pos_hits[:2] or neg_hits[:2]
    return [{
        "text": (
            f"Management tone reads {lean} on a Loughran-McDonald-style lexicon "
            f"({len(pos_hits)} positive vs {len(neg_hits)} negative sentiment lines). "
            "Baseline signal, not a verdict."
        ),
        "citations": sorted(set(cites)) or [1],
    }]


def _extract_quotes(lines: list[str]) -> list[dict]:
    out: list[dict] = []
    for i, text in enumerate(lines, start=1):
        speaker = re.match(r"^([A-Z][a-zA-Z.\- ]{2,40}?)\s*[:\-]", text)
        words = set(re.findall(r"[a-z]+", text.lower()))
        strong = words & (LM_POSITIVE | LM_NEGATIVE)
        if speaker and strong and len(text) < 240:
            body = text[speaker.end():].strip()
            out.append({"text": _clip(body), "speaker": speaker.group(1).strip(), "citations": [i]})
            if len(out) >= 2:
                break
    return out


def _clip(text: str, limit: int = 220) -> str:
    text = text.strip()
    return text if len(text) <= limit else text[: limit - 1].rstrip() + "…"


# --- Claude backend ------------------------------------------------------------------

DEFAULT_MODEL = os.environ.get("ECA_MODEL", "claude-sonnet-5")


def _analyze_claude(transcript: Transcript, company_hint: str, quarter_hint: str, model: str | None) -> dict:
    try:
        import anthropic
    except ModuleNotFoundError as exc:  # pragma: no cover - optional dep
        raise RuntimeError(
            "Real analysis needs the anthropic SDK (`pip install anthropic`). "
            "Or run with --mock for the offline pipeline."
        ) from exc
    if not os.environ.get("ANTHROPIC_API_KEY"):
        raise RuntimeError("ANTHROPIC_API_KEY is not set. Set it, or run with --mock.")

    client = anthropic.Anthropic()
    user = prompts.build_user_prompt(company_hint, quarter_hint, transcript.numbered_markdown())
    resp = client.messages.create(
        model=model or DEFAULT_MODEL,
        max_tokens=4000,
        system=prompts.SYSTEM_PROMPT,
        messages=[{"role": "user", "content": user}],
    )
    text = "".join(block.text for block in resp.content if getattr(block, "type", "") == "text")
    return _coerce_schema(_parse_json(text))


def _parse_json(text: str) -> dict:
    text = text.strip()
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*|\s*```$", "", text, flags=re.S)
    start, end = text.find("{"), text.rfind("}")
    if start == -1 or end == -1:
        raise ValueError("Model did not return a JSON object.")
    return json.loads(text[start : end + 1])


def _coerce_schema(data: dict) -> dict:
    out = {"company": data.get("company", "Unknown Company"), "quarter": data.get("quarter", "Unknown Quarter")}
    for section in prompts.SECTIONS:
        items = data.get(section) or []
        out[section] = items if isinstance(items, list) else []
    return out

File: earnings-call-analyzer/web/src/analyze.ts
// Analysis engine. Two backends, one schema (mirrors eca/analyze.py):
//   sample : deterministic heuristic extractor (no key) — for offline demo + local dev.
//   claude : real analysis via the Anthropic Messages API when a key is provided.
// Every produced item carries real line citations so the grounding layer can verify it.

import type { Analysis, Metric, CitedClaim, Quote } from "./types.js";
import { numberedMarkdown, type Transcript } from "./ingest.js";
import { SYSTEM_PROMPT, buildUserPrompt } from "./prompts.js";

export const DEFAULT_MODEL = "claude-sonnet-5";

const LM_POSITIVE = new Set([
  "growth", "grew", "strong", "record", "improved", "improvement", "gain", "gains",
  "profit", "profitable", "outperform", "beat", "exceeded", "confident", "momentum",
  "accelerate", "expansion", "robust", "healthy", "opportunity", "progress",
]);
const LM_NEGATIVE = new Set([
  "decline", "declined", "weak", "weakness", "loss", "losses", "headwind", "headwinds",
  "pressure", "uncertain", "uncertainty", "challenge", "challenging", "slowdown",
  "miss", "missed", "risk", "risks", "soft", "softness", "adverse", "litigation",
]);

const MONEY = /\$\s?\d[\d,]*(?:\.\d+)?\s?(?:billion|million|trillion|bn|mn|b|m)?/i;
const PERCENT = /\d[\d,]*(?:\.\d+)?\s?%/;
const NUMBER = /\d[\d,]*(?:\.\d+)?/;
const MONEY_RANGE =
  /\$\s?\d[\d,]*(?:\.\d+)?\s*(?:to|-|–|and)\s*\$?\s?\d[\d,]*(?:\.\d+)?\s?(?:billion|million|trillion|bn|mn|b|m)?/i;
const PERCENT_RANGE = /\d[\d,]*(?:\.\d+)?\s*(?:to|-|–)\s*\d[\d,]*(?:\.\d+)?\s?%/;

const METRIC_KEYWORDS: [string, RegExp][] = [
  ["EPS", /\b(eps|earnings per share|diluted (?:eps|earnings))\b/i],
  ["Revenue", /\b(revenue|net sales|total sales|top line)\b/i],
  ["Net income", /\b(net income|net earnings|net loss)\b/i],
  ["Operating margin", /\b(operating margin|gross margin|operating income)\b/i],
  ["Free cash flow", /\b(free cash flow|operating cash flow|cash flow)\b/i],
  ["Guidance", /\b(guidance|outlook|expect|forecast|full[- ]year|fiscal)\b/i],
  ["Buyback / dividend", /\b(buyback|repurchase|dividend|return of capital)\b/i],
];
const GUIDANCE_RE = /\b(guidance|outlook|expect|anticipate|forecast|full[- ]year|next quarter|fiscal \d{4})\b/i;
const RISK_RE = /\b(risk|headwind|uncertain|challenge|pressure|macro|foreign exchange|fx|slowdown|softness)\b/i;
const QA_MARKER = /\b(q&a|question[- ]and[- ]answer|first question|analyst)\b/i;
const COMPANY_RE =
  /((?:[A-Z][A-Za-z0-9.&'’]+,?\s+){1,5}(?:Corporation|Incorporated|Company|Holdings|Group|Corp|Inc|Ltd|PLC|Co|N\.V)\.?)/;

export interface AnalyzeOptions {
  transcript: Transcript;
  companyHint?: string;
  quarterHint?: string;
  apiKey?: string;
  model?: string;
}

export async function analyze(
  opts: AnalyzeOptions,
): Promise<{ analysis: Analysis; backend: "claude" | "sample"; model: string }> {
  if (opts.apiKey) {
    const model = opts.model || DEFAULT_MODEL;
    const analysis = await analyzeClaude(opts.transcript, opts.companyHint ?? "", opts.quarterHint ?? "", opts.apiKey, model);
    return { analysis, backend: "claude", model };
  }
  return { analysis: analyzeSample(opts.transcript, opts.companyHint ?? "", opts.quarterHint ?? ""), backend: "sample", model: "sample" };
}

// --- Sample (deterministic) backend --------------------------------------------------

export function analyzeSample(t: Transcript, companyHint: string, quarterHint: string): Analysis {
  const lines = t.lines;
  const company = companyHint || guessCompany(lines);
  const quarter = quarterHint || guessQuarter(lines);
  const key_metrics = extractMetrics(lines);
  const guidance = extractMatches(lines, GUIDANCE_RE, 4);
  const risks = extractMatches(lines, RISK_RE, 4);
  const qa_highlights = extractQA(lines);
  const management_tone = extractTone(lines);
  const notable_quotes = extractQuotes(lines);

  const executive_summary: CitedClaim[] = [];
  for (const m of key_metrics.slice(0, 3)) {
    executive_summary.push({ text: `${m.metric} reported at ${m.value}.`, citations: m.citations });
  }
  if (management_tone.length) executive_summary.push(management_tone[0]);

  return { company, quarter, executive_summary, guidance, key_metrics, management_tone, risks, qa_highlights, notable_quotes };
}

function guessCompany(lines: string[]): string {
  for (const text of lines.slice(0, 12)) {
    const m = COMPANY_RE.exec(text);
    if (m) return m[1].trim().replace(/,$/, "");
  }
  return "Unknown Company";
}

function guessQuarter(lines: string[]): string {
  const re = /\b(Q[1-4]|first|second|third|fourth)[- ]?(?:quarter)?[, ]*(?:of\s*)?(?:fiscal\s*)?(20\d{2})\b/i;
  for (const text of lines.slice(0, 20)) {
    const m = re.exec(text);
    if (m) return `${m[1].toUpperCase()} ${m[2]}`;
  }
  return "Unknown Quarter";
}

function extractMetrics(lines: string[]): Metric[] {
  const metrics: Metric[] = [];
  const seen = new Set<string>();
  for (let i = 0; i < lines.length; i++) {
    const text = lines[i];
    if (!(MONEY.test(text) || PERCENT.test(text))) continue;
    for (const [label, kw] of METRIC_KEYWORDS) {
      if (seen.has(label)) continue;
      if (kw.test(text)) {
        const value = pickValue(text);
        if (value) {
          metrics.push({ metric: label, value, citations: [i + 1] });
          seen.add(label);
        }
        break;
      }
    }
  }
  return metrics;
}

function pickValue(text: string): string {
  const mr = MONEY_RANGE.exec(text);
  if (mr) return mr[0].trim().replace(/\s+/g, " ");
  const pr = PERCENT_RANGE.exec(text);
  if (pr) return pr[0].trim().replace(/\s+/g, " ");
  const money = MONEY.exec(text);
  const pct = PERCENT.exec(text);
  if (money && pct) return `${money[0].trim()} (${pct[0].trim()})`;
  if (money) return money[0].trim();
  if (pct) return pct[0].trim();
  const num = NUMBER.exec(text);
  return num ? num[0] : "";
}

function extractMatches(lines: string[], pattern: RegExp, limit: number): CitedClaim[] {
  const out: CitedClaim[] = [];
  for (let i = 0; i < lines.length; i++) {
    if (pattern.test(lines[i])) {
      out.push({ text: clip(lines[i]), citations: [i + 1] });
      if (out.length >= limit) break;
    }
  }
  return out;
}

function extractQA(lines: string[]): CitedClaim[] {
  let start = -1;
  for (let i = 0; i < lines.length; i++) {
    if (QA_MARKER.test(lines[i])) { start = i; break; }
  }
  if (start < 0) return [];
  const out: CitedClaim[] = [];
  for (let i = start; i < lines.length; i++) {
    const text = lines[i];
    if (/^(analyst|q:|question)/i.test(text) || PERCENT.test(text) || MONEY.test(text)) {
      out.push({ text: clip(text), citations: [i + 1] });
      if (out.length >= 4) break;
    }
  }
  return out;
}

function words(text: string): Set<string> {
  return new Set((text.toLowerCase().match(/[a-z]+/g) ?? []));
}

function extractTone(lines: string[]): CitedClaim[] {
  const pos: number[] = [];
  const neg: number[] = [];
  for (let i = 0; i < lines.length; i++) {
    const w = words(lines[i]);
    if ([...w].some((x) => LM_POSITIVE.has(x))) pos.push(i + 1);
    if ([...w].some((x) => LM_NEGATIVE.has(x))) neg.push(i + 1);
  }
  const total = pos.length + neg.length;
  if (total === 0) return [];
  const net = (pos.length - neg.length) / total;
  const lean = net > 0.15 ? "net positive" : net < -0.15 ? "net negative" : "balanced";
  const cites = [...pos.slice(0, 2), ...neg.slice(0, 2)];
  const uniq = [...new Set(cites)].sort((a, b) => a - b);
  return [{
    text:
      `Management tone reads ${lean} on a Loughran-McDonald-style lexicon ` +
      `(${pos.length} positive vs ${neg.length} negative sentiment lines). Baseline signal, not a verdict.`,
    citations: uniq.length ? uniq : [1],
  }];
}

function extractQuotes(lines: string[]): Quote[] {
  const out: Quote[] = [];
  for (let i = 0; i < lines.length; i++) {
    const text = lines[i];
    const speaker = /^([A-Z][a-zA-Z.\- ]{2,40}?)\s*[:\-]/.exec(text);
    const w = words(text);
    const strong = [...w].some((x) => LM_POSITIVE.has(x) || LM_NEGATIVE.has(x));
    if (speaker && strong && text.length < 240) {
      out.push({ text: clip(text.slice(speaker[0].length).trim()), speaker: speaker[1].trim(), citations: [i + 1] });
      if (out.length >= 2) break;
    }
  }
  return out;
}

function clip(text: string, limit = 220): string {
  text = text.trim();
  return text.length <= limit ? text : text.slice(0, limit - 1).trimEnd() + "…";
}

// --- Claude backend ------------------------------------------------------------------

async function analyzeClaude(t: Transcript, companyHint: string, quarterHint: string, apiKey: string, model: string): Promise<Analysis> {
  const user = buildUserPrompt(companyHint, quarterHint, numberedMarkdown(t));
  const resp = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-api-key": apiKey,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify({
      model,
      max_tokens: 4000,
      system: SYSTEM_PROMPT,
      messages: [{ role: "user", content: user }],
    }),
  });
  if (!resp.ok) {
    const detail = await resp.text().catch(() => "");
    throw new Error(`Anthropic API error ${resp.status}: ${detail.slice(0, 300)}`);
  }
  const data = (await resp.json()) as { content?: { type: string; text?: string }[] };
  const text = (data.content ?? []).filter((b) => b.type === "text").map((b) => b.text ?? "").join("");
  return coerceSchema(parseJson(text));
}

function parseJson(text: string): Record<string, unknown> {
  let t = text.trim();
  if (t.startsWith("```")) t = t.replace(/^```(?:json)?\s*/, "").replace(/\s*```$/, "");
  const start = t.indexOf("{");
  const end = t.lastIndexOf("}");
  if (start === -1 || end === -1) throw new Error("Model did not return a JSON object.");
  return JSON.parse(t.slice(start, end + 1));
}

function coerceSchema(data: Record<string, unknown>): Analysis {
  const arr = (k: string) => (Array.isArray(data[k]) ? (data[k] as any[]) : []);
  return {
    company: (data.company as string) || "Unknown Company",
    quarter: (data.quarter as string) || "Unknown Quarter",
    executive_summary: arr("executive_summary"),
    guidance: arr("guidance"),
    key_metrics: arr("key_metrics"),
    management_tone: arr("management_tone"),
    risks: arr("risks"),
    qa_highlights: arr("qa_highlights"),
    notable_quotes: arr("notable_quotes"),
  };
}
